cpp/src/io/convert/dlpack/cudf_dlpack.cpp

```diff
@@ -0,0 +1,247 @@
+  GDF_REQUIRE(*columns != nullptr, GDF_MEMORYMANAGER_ERROR);
+
+  gdf_size_type byte_width = gdf_dtype_size(dtype);
+  gdf_size_type length = tensor->dl_tensor.shape[0];
+  size_t bytes = length * byte_width;
+
+  int64_t col_stride = 0;
+  if (*num_columns > 1) {
+    col_stride = byte_width * length;
+    if (nullptr != tensor->dl_tensor.strides)
+      col_stride = tensor->dl_tensor.strides[1];
+  }
+
+  // copy the dl_tensor data
+  for (gdf_size_type c = 0; c < *num_columns; ++c) {
+    void *tensor_data = reinterpret_cast<void*>(
+      reinterpret_cast<uintptr_t>(tensor->dl_tensor.data) +
+      tensor->dl_tensor.byte_offset +
+      col_stride);
+
+    void* col_data = 0;
+    RMM_TRY(RMM_ALLOC(&col_data, bytes, 0));
+
+    CUDA_TRY(cudaMemcpy(col_data, tensor_data, bytes, cudaMemcpyDefault));
+
+    // construct column view
+    gdf_error status = gdf_column_view(&(*columns)[c], col_data,
+                                       nullptr, length, dtype);
+    GDF_REQUIRE(GDF_SUCCESS == status, status);
+  }
+
+  // Call the managed tensor's deleter since our "borrowing" is done
+  tensor->deleter(const_cast<DLManagedTensor*>(tensor));
+}
+
+// Convert an array of gdf_column(s) into a DLPack DLTensor
+// Supports 1D and 2D tensors (single or multiple columns)
+gdf_error gdf_to_dlpack(DLManagedTensor *tensor,
+                        gdf_column const * const * columns, 
+                        int num_columns)
+{
+  GDF_REQUIRE(columns && num_columns > 0, GDF_DATASET_EMPTY);
+
+  // first column determines datatype and number of rows
+  gdf_dtype type = columns[0]->dtype;
+  gdf_size_type num_rows = columns[0]->size;
+
+  GDF_REQUIRE(type != GDF_invalid, GDF_UNSUPPORTED_DTYPE);
+  GDF_REQUIRE(num_rows > 0, GDF_DATASET_EMPTY);
+
+  // ensure all columns are the same type and size
+  for (gdf_size_type i = 1; i < num_columns; ++i) {
+    GDF_REQUIRE(columns[i]->dtype == type, GDF_DTYPE_MISMATCH);
+    GDF_REQUIRE(columns[i]->size == num_rows, GDF_COLUMN_SIZE_MISMATCH);
+  }
+
+  tensor->dl_tensor.ndim = (num_columns > 1) ? 2 : 1;
+  
+  tensor->dl_tensor.dtype = 
+    cudf::type_dispatcher(type, gdf_dtype_to_DLDataType() );
+  GDF_REQUIRE(tensor->dl_tensor.dtype.code != 0xFF, GDF_UNSUPPORTED_DTYPE);
+    
+  tensor->dl_tensor.shape = new int64_t[tensor->dl_tensor.ndim];
+  tensor->dl_tensor.shape[0] = num_rows;
+  if (tensor->dl_tensor.ndim > 1) 
+    tensor->dl_tensor.shape[1] = num_columns;
+  tensor->dl_tensor.strides = nullptr;
+  tensor->dl_tensor.byte_offset = 0;
+  
+  CUDA_TRY( cudaGetDevice(&tensor->dl_tensor.ctx.device_id) );
+  tensor->dl_tensor.ctx.device_type = kDLGPU;
+
+  // If there is only one column, then a 1D tensor can just copy the pointer
+  // to the data in the column, and the deleter should not delete the original
+  // data. However, if there are multiple columns, we must do a copy of each
+  // column's data into the dense tensor array.
+  if (num_columns == 1) {
+    tensor->dl_tensor.data = columns[0]->data;
```

Comment:
This seems like it could lead to surprises, because you've got two entities pointing to the same data and if someone goes on after this operation and modifies the column/tensor... one or the other could be surprised that their data has changed.

It could be surprising because this behavior is only true for 1D tensors, whereas 2D tensors make a copy. So it may require users to special case the two code paths... 

However, you do ensure in the deleter that we won't try and free it twice, so at least it won't cause memory errors.

I'm not necessarily asking that you change this behavior, but rather am reminded at how one of the biggest questions we need to answer in the C++ redesign is "Who owns the memory?" and "How do we enforce that ownership?", because there is a mishmash of different schemes present in libcudf on who has the responsibility for allocating/freeing memory and whether or not data is invariant or not.

Reply:
Decided to eliminate the no-copy optimization for consistency / lack of surprises. Revisit later when we have the ability to control ownership / lifetime of memory better.